### ISSUE TITLE
Use of temporary IPv6 source address 

### ISSUE THREAD

**Issue by hzuleger** [radcli/radcli#23]
I have written a patch to use in any case the non_temporary ipv6 source address.
With this it's no longer necessary to set the source address with the bind config command in case radcli runs on a box witch uses temporary addresses as default.
You can download the patch from [1]. Currently the macro to compile this code is set unconditionally at the beginning of the source file. You can set this by an configure option or add an option to the config file like "bind non-temp".

[1](http://www.hznet.de/0001-Use-non-temporary-IPv6-source-address.patch)

**Comment by nmav** [comment]
Hi, could you provide some more context on why this is needed, and send it as a pull request?

**Comment by hzuleger** [comment]
On 11.08.2017 13:16, Nikos Mavrogiannopoulos wrote:
> Hi, could you provide some more context on why this is needed, and send
> it as a pull request?

I use radcli on a ocserv implemention running on a linux machine with

$ sysctl net.ipv6.conf.all --pattern temp
net.ipv6.conf.all.temp_prefered_lft = 86400
net.ipv6.conf.all.temp_valid_lft = 604800
net.ipv6.conf.all.use_tempaddr = 2

which means that every 24 hours a new temporary IPv6 address is
generated and this one will be used for outgoing packets by default.

On the radius server I have to identify the client via its IPv6 address
(or dns name) and the configured shared secret.

But with my defaulr config, the radiusclient takes the temporary address
as source adress and this one doesn't match with the configured one on
the radius server.

I know that I can specify the source address with the bind configuration
option, but this means that I have to configure the IPv6 address in
another config file.
With the patch this is no longer needed.

About the pull request: Sorry I'm not very familiar with Git and Github
so, to be honest I don't know how to manage this.

Best regards
 Holger

**Comment by nmav** [comment]
On the patch, it may be reasonable to detect if the system has this capabilities (with ifdef macros or configure script), and then enable USE_NON_TEMP_ADDR rather than enable it by default and fail to compile otherwise.

About pull requests, it is pretty straightforward, and you'll also see whether your patch passes the automated tests:
https://help.github.com/articles/creating-a-pull-request/

**Comment by aravindprasads** [comment]
Hi @nmav  @hzuleger  

Sorry, am unable to get the requirement of this issue. Currently, we can use "bindaddr" option in rc_add_config() API and make the radius client use it always as the source address. 

```
"I know that I can specify the source address with the bind configuration
option, but this means that I have to configure the IPv6 address in
another config file."
```
Is the rc_add_config API not sufficient ? With it, use of another file to hold IPV6 address may not be needed. Kindly provide more info if possible. Thanks for the support.

**Comment by hzuleger** [comment]
Hi,

> Sorry, am unable to get the requirement of this issue. Currently, we
> can use "bindaddr" option in rc_add_config() API and make the radius
> client use it always as the source address.
> 
> '''
> I know that I can specify the source address with the bind configuration
> option, but this means that I have to configure the IPv6 address in
> another config file.
> '''
Maybe I was a bit unclear on this topic.

What I meant is, if you have to put an IPv6 address into a config file,
than you have to change this config file in case of address re-numbering.
However, re-numbering [1] is supported in IPv6 by design and can be done
in an automated way if, and only if you do *not* put literal ipv6
addresses into config files.
So, just to set the "please use the non-tempopary address" flag in the
socket call would help in case of re-numbering.

Thank you very mutch for your support
 Holger

[1] https://tools.ietf.org/html/rfc4192

**Comment by aravindprasads** [comment]
Hi @hzuleger @nmav  

Thanks for the response. 
Am still not exactly clear of the requirement. Following is the current behavior of radius client (correct me if am wrong):
1) If the bindaddr option in config file is set to "*", the packets from client will automatically take the source address of the interface from which packet is sent. 
2) If the bindaddr option in config file is set to any specific IP address, packets will be sent using that as Source address always. 

If you need the radius-client to always use specific source-address and should not change in case of IPV6 address renumbering:
1) Provide a specific IP address for bindaddr option in config file
or
2) Use the rc_add_config() API to add the bindaddr option with specific address. 

And if you need radius-client to use newly asssigned IPv6 address after renumbering always:
1) Provide "*" for the bindaddr option in config file
or
2) Use the rc_add_config() API to add the bindaddr option with "*" value 

Kindly let me know If i am missing anything.

**Comment by hzuleger** [comment]
Hi,

> Am still not exactly clear of the requirement. Following is the current behavior of radius client (correct me if am wrong):
> 1) If the bindaddr option in config file is set to "*", the packets from client will automatically take the source address of the interface from which packet is sent. 
Yes but which source address is it?

In IPv6 you have a permanent ipv6 address and a temporary ipv6 address.
(depends on the configuration of the operating system, see below)

e.g:
$ ip -6 addr show | grep inet6 | sed 's/ *inet6 //' | grep global
2a00:34:d12:1:3042:afe:9b1a:abcf/64 scope global temporary dynamic
2a00:34:d12:1:5e26:3ef:fe85:dc33/64 scope global mngtmpaddr
noprefixroute dynamic

If the kernel parameter use_tempaddr is set to 2 (all this is linux
stuff) than each client process will prefer the temporary address of the
interface instead of the non temporary.

Here is the extract of the ip-sysctl.txt documentation:

use_tempaddr - INTEGER
	Preference for Privacy Extensions (RFC3041).
	  <= 0 : disable Privacy Extensions
	  == 1 : enable Privacy Extensions, but prefer public
	         addresses over temporary addresses.
	  >  1 : enable Privacy Extensions and prefer temporary
	         addresses over public addresses.
	Default:  0 (for most devices)
		 -1 (for point-to-point devices and loopback devices)

That means, you can *not* use the "bindaddr *" config because this IP
address changes over the time (usually every 24h, depends on the
temp_prefered_lft).

$ sysctl net.ipv6.conf.default | grep temp
net.ipv6.conf.default.temp_prefered_lft = 86400
net.ipv6.conf.default.temp_valid_lft = 604800
net.ipv6.conf.default.use_tempaddr = 2


> 2) If the bindaddr option in config file is set to any specific IP address, packets will be sent using that as Source address always. 
Yes of course, but this is the only option you have.

> If you need the radius-client to always use specific source-address and should not change in case of IPV6 address renumbering:
> 1) Provide a specific IP address for bindaddr option in config file
> or
> 2) Use the rc_add_config() API to add the bindaddr option with specific address. 
Maybe this is an option too. But again some manual change process is needed.

> And if you need radius-client to use newly asssigned IPv6 address after renumbering always:
> 1) Provide "*" for the bindaddr option in config file
> or
> 2) Use the rc_add_config() API to add the bindaddr option with "*" value 
> 
> Kindly let me know If i am missing anything. Also, if any enhancements are required aprart from the present ones. 
If the radius client set the socket option to use the non-temporary
address, than it is possible to use the "*" config in the file, and
works well.

BR
 Holger

**Comment by aravindprasads** [comment]
@hzuleger 

Thanks a lot for your time and providing a complete explanation. 

Understood your requirement. 

https://github.com/radcli/radcli/pull/44 - Is this the code change you preferred for ? 

Thanks again for the support.

**Comment by aravindprasads** [comment]
Hi @hzuleger @nmav 

I have completed the changes required to fix this issue - https://github.com/radcli/radcli/pull/44 

Kindly review the same and let me know your views.

**Comment by hzuleger** [comment]
Hi,

> I have completed the changes required to fix this issue - https://github.com/radcli/radcli/pull/44 
looks great. Thank you very mutch for your support.

The only thing I have in mind is that this works only for Linux.
BSD Unix (and Apple as well) have something similar, but a bit different.

The socket option is called "IPV6_PREFER_TEMPADDR" and it has to be set
to "0".

Instead of:

int sock_opt = IPV6_PREFER_SRC_PUBLIC;
if (setsockopt(sockfd, IPPROTO_IPV6, IPV6_ADDR_PREFERENCES,
				&sock_opt, sizeof(sock_opt)) != 0) {
	rc_log(LOG_ERR, "rc_send_server: setsockopt: %s",
			strerror(errno));
	result = ERROR_RC;
	goto cleanup;
}

the BSD setsockopt()  call will look like:

int sock_opt = 0;
if (setsockopt(sockfd, IPPROTO_IPV6, IPV6_PREFER_TEMPADDR,
				&sock_opt, sizeof(sock_opt)) != 0) {

And you have to include <netinet/in.h>


Thanks again for your work,
BR
 Holger

**Comment by aravindprasads** [comment]
Hi @hzuleger  @nmav 

Thanks a lot for the feedback. Thought, We could add the Platform on the need basis. 

I have added handlers for BSD and Apple now. 
<netinet/in.h> is included by default in "include/includes" file and didnt need separate ifdef for adding it in the include section of sendserver.c file. 

Kindly let me know if the changes are fine. Thanks for the support.

**Comment by hzuleger** [comment]
Hi,

> Thanks a lot for the feedback. Thought, We could add the Platform on the need basis. 
> 
> I have added handlers for BSD and Apple now. 
> <netinet/in.h> is included by default in "include/includes" file and didnt need separate ifdef for adding it in the include section of sendserver.c file. 
> 
> Kindly let me know if the changes are fine. Thanks for the support. 
Looks fine. Great work. Thanks a lot.

BR
 Holger

**Comment by aravindprasads** [comment]
@hzuleger 
Thanks a lot for the sustained support and your time to explain the issue in detail. 

Hi @nmav  
Kindly review the changes and let me know your views.

**Comment by nmav** [comment]
This was addressed in 1.2.11.
